ivy.lexsort added

nadidebeyza opened the pull request:
Implementing a NumPy frontend function
Closes #7126

nadidebeyza commented:
- [ ] #7126

Daniel4078 commented:
Closes #7126

nadidebeyza commented:
> Can you try remove your change to the .idea folder's content? As they are config files for the pycharm ide I think.

Done.

Daniel4078 commented:
Wait, I just realised, you are supposed to create a experimental function lexsort in the experimental API including all the backends,  right? Then you should have put this function's frontend inside ivy/functional/experimental folder and its backends in ivy/functional/backends/[backend framework]/experimental folder.

nadidebeyza commented:
> Wait, I just realised, you are supposed to create a experimental function lexsort in the experimental API including all the backends, right? Then you should have put this function's frontend inside ivy/functional/experimental folder and its backends in ivy/functional/backends/[backend framework]/experimental folder.

@Daniel4078 I was thinking to implement the frontend function under this PR. And then I was going to create another PR for experimental API including all the backends. Should I complete both under this PR?

Daniel4078 commented:
> > Wait, I just realised, you are supposed to create a experimental function lexsort in the experimental API including all the backends, right? Then you should have put this function's frontend inside ivy/functional/experimental folder and its backends in ivy/functional/backends/[backend framework]/experimental folder.
> 
> @Daniel4078 I was thinking to implement the frontend function under this PR. And then I was going to create another PR for experimental API including all the backends. Should I complete both under this PR?

I looked at the other pull request you closed and the reopened and I assume you are doing the numpy frontend there. But I think you can complete two tasks together in one pull request.

Daniel4078 commented on ivy/array/experimental/sorting.py:
```diff
@@ -42,3 +42,16 @@ def msort(
+) -> ivy.Array:
+    """
+    """
+    return ivy.lexsort(self._data, axis=axis, out=out)
```
What you did here is let the function call itself, as the tests shows that RecursionError: maximum recursion depth exceeded.

Daniel4078 replied:
And to be more precise, it is not that the problem is here, it is that this call targets another function that will call itself and form a loop.

Daniel4078 replied:
for instance, look at this chain of error: E   ivy.exceptions.IvyBackendException: tensorflow: lexsort: tensorflow: lexsort: tensorflow: lexsort: tensorflow: lexsort: tensorflow: lexsort: tensorflow: lexsort: tensorflow: lexsort: tensorflow: nested_map: tensorflow: nested_map: tensorflow: default: tensorflow: exists: maximum recursion depth exceeded

Daniel4078 replied:
it all begins after this step happened: ivy/functional/ivy/sorting.py:259: in lexsort
    return ivy.current_backend(keys).lexsort(

Daniel4078 replied:
Which then seems to call back to the frontend function

nadidebeyza replied:
I fixed the indentation error about the "self", it is not recognized because of this. I know still I have a recursion error also, I am working on it

Daniel4078 reviewed (changes requested):
I found a strange error when I looked through your test for the experimental function.

Daniel4078 commented on ivy/functional/backends/numpy/experimental/sorting.py:
```diff
@@ -11,3 +11,14 @@ def msort(
+
+
+# lexsort
+def lexsort(
```
So I suppose you should have one of this lexsort definition in each of the backends/[frameworks]/experimencal/sorting.py file

nadidebeyza replied:
I added lexsort definition in each of the backends but it is not completed because for instance tensorflow has no tf.lexsort. I will implement also lexsort functions in ivy>functional>backends>[frameworks]>sorting.py

Daniel4078 replied:
As tensorflow does not have a lexsort for you to use, why not try to recreate the expected behavior through a combination of multiple tensorflow functions? Like adding some extra steps after using a function of tensorflow that does a similar job like lexsort?

nadidebeyza replied:
I am thinking this is the best case for lexsort torch implementation compared to the other cases I tried. But I am getting an error(AssertionError: [0 1] != 0)😬​, do you have any suggestions for me?

Daniel4078 replied:
Can I have a more detailed example of that assertion error?

nadidebeyza replied:
Yes, you are right, I am sorry. This is the all [test results of failed tests](https://drive.google.com/file/d/14THp9dnPYcrdYcoyoHKfBGfFqn8Petme/view?usp=sharing). This is 16 pages document. I marked the assertion error part(13.,14. pgs).

nadidebeyza replied:
I also rarely did well for the torch test, I don't think I've taken a probabilistic step. I'm wrong about one thing, that's for sure

Daniel4078 replied:
At first glance, it seems like your function returns a list of two numbers while the real counter part from numpy returns only 0. Then I searched about how the lexsort behave in numpy and found that your test does not provide the correct form of input it seems. https://numpy.org/doc/stable/reference/generated/numpy.lexsort.html

Daniel4078 replied:
Just to make sure, the ground truth value given by numpy's lexsort(np.lexsort([0, 0]) = 0) is the expected behavior.

nadidebeyza replied:
Wow I am stuck, I didn't expect this. I have a few scenarios that run well and these do not run the ivy tests. I have just committed another one and the error about "tuple index out of range". This is the all [test results of failed tests](https://drive.google.com/file/d/1a0kQwqESlMwhJe1_zFD6mNsC6bnUG_ze/view?usp=sharing). This is 23 pages document. I marked the out-of-range error part(pg 22).

nadidebeyza replied:
I know it looks like bad code and it is. According to my first commit about lexsort implementation in the torch backend, I thought it was clean and effective code to implement the function using the torch.unique but it failed. [torch.unique](https://pytorch.org/docs/stable/generated/torch.unique.html) returns unique elements in the tensor and can return reverse indices of them is optional. This time, I wanted to get reverse indices manually for loops. So I would save the function from the assertion error by only returning a value

nadidebeyza replied:
I am getting a "tuple index out of range" error while for example the test tensor equal to torch.Tensor([0]). I am confused because for lexsort it already must take multiple sorting keys, right?

nadidebeyza replied:
At this point, I think I have to make a decision. I was planning to complete torch.lexsort() and then tensorflow.lexsort(). What do you think should I continue to work on torch.lexsort() or should I move to the tensorflow.lexsort()?

Daniel4078 replied:
The lexsort requires a (k, N) array or tuple containing k (N,)-shaped sequences and should not work in other cases. I think this is the sourse of the out of range error you meet: 
![image](https://user-images.githubusercontent.com/45633544/205482204-06a6b77c-de09-4d54-8ec1-36627d58e0dd.png)

Daniel4078 replied:
You should first implement a way to check if the input have the correct shape by replicate what numpy will do in that case first.

nadidebeyza replied:
Now I get only one value from the torch.unique() function. 

I applied what you said and I found the exact reason for the error is numpy.lexsort() function returns for example: [0, 0] = [0], [1, 5, 9] = [0]. So if the input is a vector result will be 0.

In my function inverse_indices returns a tensor (same shape as input) representing the indices for where elements in the original input map to in the output. It means torch.argsort(inverse_indices, dim=axis)  returns [0, 0] = [1, 5, 9], [2, 1, 0]

Daniel4078 commented on ivy/functional/backends/numpy/sorting.py:
```diff
@@ -73,3 +73,15 @@ def searchsorted(
     return ret.astype(ret_dtype)
+
+
+def lexsort(
```
Also, are you sure that you need to implement the whole thing again not in the experimental folder? That seems redundent and your frontend function can just call your ivy.experimental function.

nadidebeyza replied:
I am getting an error when I don't implement it, so I will keep this in the current situation and after resolving other things I will evaluate this again. Would this be okay for you?

Daniel4078 commented:
@ahmedo42 I need your suggestions about what to do in this case.

nadidebeyza commented:
> Just as a reminder, yesterday there was a big change to the code base that fix the failing tests due to the previously mensioned handler refactor which changes how tests should be written.

Aaa okay I understand. Actually I was going to reply you about my last commit. I understood your last post and took it into consideration, thank you. But I am working on it from yesterday also for a new one because it was not the solution just I wanted to see how is going something

Daniel4078 commented on ivy/functional/backends/tensorflow/experimental/sorting.py:
```diff
@@ -11,3 +11,14 @@ def msort(
+    axis: int = -1,
+    out: Optional[Union[tf.Tensor, tf.Variable]] = None,
+) -> Union[tf.Tensor, tf.Variable]:
+    return tf.lexsort(x, axis)
```
By the way, do not forget there is no lexsort in tensorflow. Please try a compositional approach.

Daniel4078 commented:
Next time if you have a question or something, please request a review from me, so I can get more conveniently notified. Thank you.

nadidebeyza commented:
> Just as a reminder, yesterday there was a big change to the code base that fix the failing tests due to the previously mensioned handler refactor which changes how tests should be written.

I will do so, thank you🌱

nadidebeyza commented:
> Next time if you have a question or something, please request a review from me, so I can get more conveniently notified. Thank you.

I will do so, thank you🌱

Daniel4078 commented:
How is your current progress? Do you need help?

nadidebeyza commented:
> > How is your current progress? Do you need help?
> 
> Sorry for my late reply. I have already implemented the solution I told you about last. Surprisingly when I run my code three options can occur, it changes according to test cases.
> 
> 1. Test completed successfully.
> 2. Getting "assertion error".
> 3. Getting "uniquedim is not implemented for Half".

The same solution run well with the same input separately from the ivy code

![output](https://user-images.githubusercontent.com/48105379/206760265-a1ea6bec-a4c0-4ee8-8c58-292dc549f12d.png)

nadidebeyza commented:
> > > How is your current progress? Do you need help?
> > 
> > 
> > Sorry for my late reply. I have already implemented the solution I told you about last. Surprisingly when I run my code three options can occur, it changes according to test cases.
> > 
> > 1. Test completed successfully.
> > 2. Getting "assertion error".
> > 3. Getting "uniquedim is not implemented for Half".
> 
> The same solution run well with the same input separately from the ivy code
> 
> ![output](https://user-images.githubusercontent.com/48105379/206760265-a1ea6bec-a4c0-4ee8-8c58-292dc549f12d.png)

Also about the error  'half' is not implemented, I was searching the same kind of error on the discord channel and I found that is probably because of the backend does not support HalfTensor data type as input for the function

nadidebeyza commented:
> > > How is your current progress? Do you need help?
> > 
> > 
> > Sorry for my late reply. I have already implemented the solution I told you about last. Surprisingly when I run my code three options can occur, it changes according to test cases.
> > 
> > 1. Test completed successfully.
> > 2. Getting "assertion error".
> > 3. Getting "uniquedim is not implemented for Half".
> 
> The same solution run well with the same input separately from the ivy code
> 
> ![output](https://user-images.githubusercontent.com/48105379/206760265-a1ea6bec-a4c0-4ee8-8c58-292dc549f12d.png)

Also about the error  'half' is not implemented, I was searching the same kind of error on the discord channel and I found that is probably because of the backend does not support HalfTensor data type as input for the function

Daniel4078 commented:
That's unfortunate. Hope it gets resolved soon. So do you think your implementations works now?

nadidebeyza commented:
@Daniel4078 Hi again Daniel, I decided to complete my implementation PR as only the front-end compositionally without adding an experimental API function. Experimental-api part completed for numpy and jax. If is it possible to merge only them without torch and tensorflow I want also merge them too. Can you guide me on how to proceed? Should I open a new PR or etc.?

nadidebeyza commented:
> @Daniel4078 Hi again Daniel, I decided to complete my implementation PR as only the front-end compositionally without adding an experimental API function. Experimental-api part completed for numpy and jax. If is it possible to merge only them without torch and tensorflow I want also merge them too. Can you guide me on how to proceed? Should I open a new PR or etc.?

@Daniel4078 @ahmedo42 I wanted to kindly remind you my message...

Daniel4078 commented:
I think you do not need to create a new PR, just remove the unnecessary changes and I have removed those extra labels for you. Thank you for your notice, I was pretty busy last week dealing with cleaning the history of the ivy codebase.

nadidebeyza commented:
> I think you do not need to create a new PR, just remove the unnecessary changes and I have removed those extra labels for you. Thank you for your notice, I was pretty busy last week dealing with cleaning the history of the ivy codebase.

I understand no problem, thank you for answering now. I think I have removed the unnecessary changes carefully but still I am not sure

Daniel4078 commented on ivy/functional/backends/jax/experimental/sorting.py:
```diff
@@ -14,3 +14,14 @@ def msort(
     return jnp.msort(a)
+
+
+# lexsort
```
like these changes in the experiment folders I think

Daniel4078 commented:
well, the experimental functions need to have all the backend working together, so you know, you can save those correct implementations for the future.

nadidebeyza commented:
> well, the experimental functions need to have all the backend working together, so you know, you can save those correct implementations for the future.

I understand. I removed them all and numpy frontend test runs well for the lexsort function

Daniel4078 commented:
I just found that there is this strange error with your numpy test:
tensorflow: nested_map: maximum recursion depth exceeded in comparison
and the test seems to be stuck in an infinite loop or something. Please have a look.

nadidebeyza commented:
> I just found that there is this strange error with your numpy test: tensorflow: nested_map: maximum recursion depth exceeded in comparison and the test seems to be stuck in an infinite loop or something. Please have a look.

Sorry for my late reply, I was hoping I had time to fix the error so I didn't get back to you that day but I couldn't due to my final week. It's hard for me to see the error as pycharm passed the test in console without any problems, I'm working on it now. I wanted to let you know

Daniel4078 commented:
Take your time, do not rush as there are many changes to the test functions recently so it is hard to get tests passing consistently, by the way, do not forget to pull from the master from time to time.

ivy-seed commented:
This PR has been labelled as stale because it has been inactive for more than 7 days. If you would like to continue working on this PR, then please add another comment or this PR will be closed in 7 days.

nadidebeyza commented:
I am still working on it

Daniel4078 commented:
well, that is a bot you know. you can just make a commit from time to time.

nadidebeyza commented:
UPDATE COMMENT!
- Good news: I will be graduated in 1.5 weeks, I am determined to complete my task.
- Bad news: No bad news.

Daniel4078 commented:
How is it going? Please do not forget to keep your branch updated.

Daniel4078 commented:
do not forget to keep your branch updated with the master branch.

nadidebeyza commented:
> do not forget to keep your branch updated with the master branch.

Sorry I didn't see your comment. I have just updated, thank you. Now I am ready to work on it again

Daniel4078 commented on ivy/functional/ivy/sorting.py:
```diff
@@ -234,6 +234,27 @@ def sort(
+    stable: bool = True,
+    out: Optional[ivy.Array] = None,
+) -> ivy.Array:
+    """
```
just to confirm, does the contribution guide or the todo list issue mension that you need to filling the docstring for the function as well?

nadidebeyza replied:
I didn't because docstring was not requested. I saw a separate tasks for docstrings

Daniel4078 commented on ivy_tests/test_ivy/test_frontends/test_numpy/test_sorting_searching_counting/test_sorting.py:
```diff
@@ -129,3 +129,35 @@ def test_numpy_sort_complex(
+    ),
+    test_with_out=st.just(False),
+)
+def test_numpy_lexsort(
```
the test still fails due to this RecursionError: maximum recursion depth exceeded. Do you have any idea why?

nadidebeyza replied:
My idea is the function has no backend implementation and so that I need to implement that in
-ivy/functional/backends/numpy/sorting.py and 
-ivy/functional/ivy/sorting.py 

When I compare with other completed frontend function implementations they didnt implemented backend for their functions because they have implemented frontend functions that have already implemented backend.

When I implemeted also the backend for numpy them the test runs well with numpy but I get an error with the other libraries. Since I did not implement it for other backends, I think that it always gets stuck in the same rejection response in the test and enters an endless loop. Is this a bug that needs to be fixed? or Is this a fixable error in my task?

Daniel4078 replied:
So do you know how or want to implement the backend for all other backends?

nadidebeyza replied:
Actually, it means I am at the beginning of everything. This is what I gave up before. Can't I just make numpy frontend and complete it?

Daniel4078 replied:
I guess you can as someone else can fill in the missing thing later?

nadidebeyza replied:
I think this is the version. For example -just for try- when I add backend implementation for jax, numpy frontend test runs well for jax too without any infinite loop. But without jax backend implementation it goes into infinite loop for jax too

Daniel4078 replied:
that sort of make sense. so maybe add a comment to the code stating that it need all backends filled in? And then fix the easy problems before I maybe just merge this?

nadidebeyza replied:
I added the comment above the frontend test function. Anyone working on it can easily reach it. What are the other easy problems?

Daniel4078 commented on ivy/functional/frontends/numpy/sorting_searching_counting/sorting.py:
```diff
@@ -28,3 +28,13 @@ def msort(a):
+    *,
+    axis=-1
+):
+    return ivy.lexsort(keys, axis=axis)
```
please do not forget about this one

nadidebeyza commented:
> Looks good enough.

Thank you for your patience and support, I learned a lot😊​💫​
